include/rapidcheck/state/gen/ParallelCommands.hpp

```diff
@@ -0,0 +1,341 @@
+#pragma once
+
+#include "rapidcheck/state/gen/Commands.h"
+
+namespace rc {
+namespace state {
+namespace gen {
+
+namespace detail {
+template <typename Cmd, typename GenFunc>
+class ParallelCommandsGen {
+public:
+  using CmdSP = std::shared_ptr<const Cmd>;
+  using Model = typename Cmd::Model;
+  using Sut = typename Cmd::Sut;
+
+  template <typename ModelArg, typename GenFuncArg>
+  ParallelCommandsGen(ModelArg &&initialState, GenFuncArg &&genFunc)
+      : m_initialState(std::forward<ModelArg>(initialState))
+      , m_genFunc(std::forward<GenFuncArg>(genFunc)) {}
+
+  Shrinkable<ParallelCommands<Cmd>> operator()(const Random &random,
+                                               int size) const {
+    return generateCommands(random, size);
+  }
+
+private:
+  struct CommandEntry {
+    CommandEntry(Random &&aRandom, Shrinkable<CmdSP> &&aShrinkable)
+        : random(std::move(aRandom))
+        , shrinkable(std::move(aShrinkable)) {}
+
```

Comment:
I agree that the latter option is most likely the right call. Since it's already a requirement that any interleaving should be valid, that sort of implies that the branches are independent anyway and thus having independent state sequences at generation time also makes sense.

But I think you have to elaborate on the predicate idea because I don't really understand :)